ludwig/hyperopt.py

```diff
@@ -0,0 +1,512 @@
+        random_seed=default_random_seed,
+        debug=False,
+        **kwargs,
+):
+    # check for model_definition and model_definition_file
+    if model_definition is None and model_definition_file is None:
+        raise ValueError(
+            "Either model_definition of model_definition_file have to be"
+            "not None to initialize a LudwigModel"
+        )
+    if model_definition is not None and model_definition_file is not None:
+        raise ValueError(
+            "Only one between model_definition and "
+            "model_definition_file can be provided"
+        )
+
+    # merge with default model definition to set defaults
+    if model_definition_file is not None:
+        with open(model_definition_file, "r") as def_file:
+            model_definition = merge_with_defaults(yaml.safe_load(def_file))
+    else:
+        model_definition = merge_with_defaults(model_definition)
+
+    if HYPEROPT not in model_definition:
+        raise ValueError("Hyperopt Section not present in Model Definition")
+
+    hyperopt_params = model_definition["hyperopt"]
+    update_hyperopt_params_with_defaults(hyperopt_params)
+
+    strategy = hyperopt_params["strategy"]
+    executor = hyperopt_params["executor"]
+    parameters = hyperopt_params["parameters"]
+
+    hyperopt_strategy = HyperoptStrategy(strategy, parameters)
+
+    hyperopt_executor = HyperoptExecutor(hyperopt_strategy, executor)
+
+    results = hyperopt_executor.execute(
+        model_definition,
+        data_df=None,
+        data_train_df=None,
+        data_validation_df=None,
+        data_test_df=None,
+        data_csv=None,
+        data_train_csv=None,
+        data_validation_csv=None,
+        data_test_csv=None,
+        data_hdf5=None,
+        data_train_hdf5=None,
+        data_validation_hdf5=None,
+        data_test_hdf5=None,
+        train_set_metadata_json=None,
+        experiment_name="hyperopt",
+        model_name="run",
+        model_load_path=None,
+        model_resume_path=None,
+        skip_save_training_description=False,
+        skip_save_training_statistics=False,
+        skip_save_model=False,
+        skip_save_progress=False,
+        skip_save_log=False,
+        skip_save_processed_input=False,
+        skip_save_unprocessed_output=False,
+        skip_save_test_predictions=False,
+        skip_save_test_statistics=False,
+        output_directory="results",
+        gpus=None,
+        gpu_fraction=1.0,
+        use_horovod=False,
+        random_seed=default_random_seed,
+        debug=False,
+        **kwargs,)
+
+    return results
+
+
+def update_hyperopt_params_with_defaults(hyperopt_params):
+
+    set_default_value(hyperopt_params, STRATEGY, {})
+    set_default_value(hyperopt_params, EXECUTOR, {})
+
+    set_default_values(
+        hyperopt_params[STRATEGY],
+        {
+            "type": "random",
+            "num_samples": 12
+        }
+    )
+
+    if hyperopt_params[STRATEGY]["type"] == "grid":
```

Comment:
fr now this is ok, later we can change it so that each HyperoptStrategy contains its default parameters and this function reads the defaults from there.